### PR TITLE
NEW: Migrate NVIDIA redist feedstocks to use cf-nvidia-tools AGAIN

### PR DESCRIPTION
**Pull request by beckermr**
This PR is ready for us to fix some initialization issue in this migrator class. I will push changes to the test suite to catch this bug too.

Reverts regro/cf-scripts#3882

### CODE REVIEW

**Comment by beckermr** [comment]
sorry for the trouble @carterbox @jakirkham - I may be able to fix this myself as well but need to move to another task right now.

**Comment by carterbox** [comment]
I am looking to try and figure out what init step is missing. I probably won't have permission to commit to this branch.

**Comment by beckermr** [comment]
I think the commit I just pushed will do the trick. As I implied above a bit, or at least mean to, the class hierarchy here has evolved organically and so acts in weird ways as different APIs and conventions have been layered on top of each other. It needs fixing, but that is a tall order at the moment. 

I need to add a test to this PR to try and ensure the effective graph is always set for all migrators. 

I hesitate to put it directly in the parent init because sometimes the child classes need to do their own things before the effective graph is created.

**Comment by beckermr** [comment]
Another option is to put it in the parent init and then have child classes who need overrides to send None to the parent init. That is probably the right way to go for now to avoid this in the future.

**Comment by carterbox** [comment]
Thanks for the patch. If this is the fix, I suggest adding the following to the docstring for Migrator:

```
class Migrator:
    """Base class for Migrators

    Inheritors
    ----------
    Subclasses of Migrator should have at least the following in their __init__ function:
    
    ```python
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._reset_effective_graph()
    ```
    
    """

```

**Comment by carterbox** [comment]
Tests failing with:

```
  Unable to find image 'mongo:latest' locally
  docker: Error response from daemon: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit.
  See 'docker run --help'.
  Error starting MongoDB Docker container
```

**Comment by beckermr** [comment]
pre-commit.ci autofix

**Comment by jakirkham** [comment]
Thanks Matt and Daniel! 🙏

Looks like CI is now passing.

What are the next steps? Sounds like add a test? Anything else?

Also please let us know if there are things you would like us to help with

**Comment by beckermr** [comment]
Yep we should add a test. I invited @carterbox to regro and the bot team, so he should be able to more easily help maintain. 

Anyone should feel free to push to this pr to add a test. Otherwise I'll do it tomorrow morning likely.

**Comment by beckermr** [comment]
It turns out there were a few more bugs and that writing a test for this edge case is really hard because we need the graph data around. 

I am going to merge this PR and baby sit things. I have opened an issue (#3888) which gets at the heart of what is happening. Once we address that, it should be a lot easier to add migrator classes by hand.

**Comment by beckermr** [comment]
@carterbox I just noticed there are not any tests of the parsing logic of the migrator on an example recipe. Can you add a test of that to this PR?

**Comment by carterbox** [comment]
Test files are now restored, but only if the test passes. If the test fails, they are not restored so that the developer can see what happened.

File: tests/test_migrator_nvtools.py [hunk 1]
```diff
@@ -18,24 +18,48 @@
...
+            shallow=False,
+        )
+        and filecmp.cmp(
+            os.path.join(mock_recipe_dir, "..", "conda-forge.yml"),
+            os.path.join(mock_recipe_dir_ref, "..", "conda-forge.yml"),
+            shallow=False,
+        )
     )
+    if success:
+        # Files only restored on success, so that you can see why the comparison failed
```
[the remark this answers is not in the document]
Comment (carterbox) [reply tests/test_migrator_nvtools.py:60]: I have used try...finally to always reset the input files for the test

**Comment by beckermr** [comment]
thank you @carterbox!
